recommendations/local-resource-conservation.md

```diff
@@ -0,0 +1,334 @@
+
+We define the following parameters: 
+* `height_added`: the block height at which the HTLC was irrevocably committed 
+  to by the local node.
+
+We define the `outstanding_risk` of in-flight HTLCs: 
+* `outstanding_risk` = `fees` * ((`cltv_expiry` - `height_added`) * 10 * 60) / `resolution_period`
+
+We define the `in_flight_htlc_risk` for an incoming channel sending 
+`update_add_htlc`:
+* `in_flight_htlc_risk` = `outstanding_risk(proposed update_add_htlc)`
+* for each HTLC originating from the incoming channel:
+  * if `endorsed` is non-zero: 
+    * `in_flight_htlc_risk` += `outstanding_risk(HTLC)`
+
+##### Rationale
+In flight HTLC are included in reputation scoring to account for sudden changes
+in a peer's behavior. Even when sufficient reputation is obtained, each HTLC 
+choosing to take advantage of that reputation is treated as if it will be used 
+to inflict maximum damage. The expiry height of each incoming in flight HTLC is 
+considered so that risk is directly related to the amount of time the HTLC 
+could be held in the channel. Ten minute blocks are assumed for simplicity.
+
+## Resource Bucketing
+When making the decision to forward a HTLC on its outgoing channel, a node 
+MAY choose to limit its exposure to HTLCs that put it at risk of a denial of 
+service attack.
+
+We define the following parameters:
+* `protected_slot_count`: defines the number of HTLC slots that are reserved 
+  for endorsed HTLCs from peers with sufficient reputation (default: 0.5 * 
+  remote peer's `max_accepted_htlcs`).
+* `protected_liquidity_portion`: defines the portion of liquidity that is 
+  reserved for endorsed HTLCs from peers with sufficient reputation (default: 
+  0.5).
+
+A node implementing resource bucketing limits exposure on its outgoing channel:
+* MUST choose `protected_slot_count` <= the remote channel peer's 
+  `max_accepted_htlcs`.
+* MUST choose `protected_liquidity_portion` in [0;1].
+
+For each `update_add_htlc` proposed by an incoming channel:
+* If `endorsed` is non-zero AND the incoming channel has sufficient local 
+  reputation for the HTLC (see [Recommendations for Reputation Scoring](#recommendations-for-reputation-scoring)):
+  * SHOULD forward the HTLC as usual.
+  * SHOULD set `endorsed` to 1 in the outgoing `update_add_htlc`.
+* Otherwise: 
+  * SHOULD reduce the remote peer's `max_accepted_htlcs` by 
+    `protected_slot_count` for the purposes of the proposed HTLC.
+  * SHOULD reduce the `max_htlc_value_in_flight` by 
+    `protected_liquidity_portion` * `max_htlc_value_in_flight`.
+  * SHOULD set `endorsed` to `0` in the outgoing `update_add_htlc`.
+
+## Implementation Notes
+
+### Decaying Average
+Rolling windows specified in this write up may be implemented as a decaying 
+average to minimize the amount of data that needs to be stored per-channel. In 
+flight HTLCs can be accounted for separately to this calculation, as the node 
+will already have data for these HTLCs available.
```

Comment:
> Rolling windows specified in this write up may be implemented as a decaying 
> average to minimize the amount of data that needs to be stored per-channel. In 
> flight HTLCs can be accounted for separately to this calculation, as the node 
> will already have data for these HTLCs available.

Minor - I believe the overall "Local Resource Conversartion" proposal would gain to
have the implementation notes dried up in its own document or blip, including
some magic values that are referenced in other subsections (e.g the 10 for the
`incoming_channel_multiplier` defining the rolling window).

Without thoughts really on the decaying average, there could be implementation
alternative such as taking all the HTLCs points since the channel opening and
periodically re-evaluating their score according to the on-chain fees, as one
can see in the blocks, or the total HTLC forwarding traffic that has been through
the target node. Just ideas, more to note the range of rolling windows algorithm 
that could be experimented with.